File: src/pocketcloud/cloud/provider/database/DatabaseQueries.php
<?php

namespace pocketcloud\cloud\provider\database;

use pocketcloud\cloud\template\TemplateHelper;
use pocketcloud\cloud\util\enum\ParameterEnumTrait;
use r3pt1s\mysql\query\QueryBuilder;

/**
 * @method static QueryBuilder createTables()
 * @method static QueryBuilder addTemplate(array $data)
 * @method static QueryBuilder removeTemplate(string $name)
 * @method static QueryBuilder editTemplate(string $name, array $newData)
 * @method static QueryBuilder getTemplate(string $name)
 * @method static QueryBuilder checkTemplate(string $name)
 * @method static QueryBuilder getTemplates()
 * @method static QueryBuilder addServerGroup(array $data)
 * @method static QueryBuilder removeServerGroup(string $name)
 * @method static QueryBuilder editServerGroup(string $name, array $newData)
 * @method static QueryBuilder getServerGroup(string $name)
 * @method static QueryBuilder checkServerGroup(string $name)
 * @method static QueryBuilder getServerGroups()
 * @method static QueryBuilder setModuleState(string $module, bool $enabled)
 * @method static QueryBuilder getModuleState(string $module)
 * @method static QueryBuilder enablePlayerNotifications(string $player)
 * @method static QueryBuilder disablePlayerNotifications(string $player)
 * @method static QueryBuilder hasNotificationsEnabled(string $player)
 * @method static QueryBuilder addToWhitelist(string $player)
 * @method static QueryBuilder removeFromWhitelist(string $player)
 * @method static QueryBuilder isOnWhitelist(string $player)
 * @method static QueryBuilder getWhitelist()
 */
final class DatabaseQueries {
    use ParameterEnumTrait;

    protected static function init(): void {
        self::register("createTables", function (): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->create([
                    "name" => "VARCHAR(50) PRIMARY KEY",
                    "lobby" => "BOOL",
                    "maintenance" => "BOOL",
                    "static" => "BOOL",
                    "maxPlayerCount" => "INTEGER",
                    "minServerCount" => "INTEGER",
                    "maxServerCount" => "INTEGER",
                    "startNewPercentage" => "FLOAT",
                    "autoStart" => "BOOL",
                    "templateType" => "VARCHAR(10)"
                ])
                ->changeTable(DatabaseTables::SERVER_GROUPS)
                    ->create([
                        "name" => "VARCHAR(50) PRIMARY KEY",
                        "templates" => "TEXT"
                    ])
                ->changeTable(DatabaseTables::MODULES)
                    ->create([
                        "module" => "VARCHAR(100) PRIMARY KEY",
                        "enabled" => "BOOL"
                    ])
                ->changeTable(DatabaseTables::NOTIFICATIONS)
                    ->create([
                        "player" => "VARCHAR(16) PRIMARY KEY"
                    ])
                ->changeTable(DatabaseTables::MAINTENANCE_LIST)
                    ->create([
                        "player" => "VARCHAR(16) PRIMARY KEY"
                    ]);
        });

        self::register("addTemplate", function (array $data): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->insert($data);
        });

        self::register("removeTemplate", function (string $name): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->delete(["name" => $name]);
        });

        self::register("editTemplate", function (string $name, array $newData): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->update($newData, ["name" => $name]);
        });

        self::register("getTemplate", function (string $name): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->get(TemplateHelper::KEYS, ["name" => $name]);
        });

        self::register("checkTemplate", function (string $name): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->has(["name" => $name]);
        });

        self::register("getTemplates", function (): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->select(TemplateHelper::KEYS, "*");
        });

        self::register("addServerGroup", function (array $data): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::SERVER_GROUPS)
                ->insert($data);
        });

        self::register("removeServerGroup", function (string $name): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::SERVER_GROUPS)
                ->delete(["name" => $name]);
        });

        self::register("editServerGroup", function (string $name, array $newData): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::SERVER_GROUPS)
                ->update($newData, ["name" => $name]);
        });

        self::register("getServerGroup", function (string $name): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::SERVER_GROUPS)
                ->get(["name", "templates"], ["name" => $name]);
        });

        self::register("checkServerGroup", function (string $name): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::SERVER_GROUPS)
                ->has(["name" => $name]);
        });

        self::register("getServerGroups", function (): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::TEMPLATES)
                ->select(TemplateHelper::KEYS, "*");
        });

        self::register("setModuleState", function (string $module, bool $enabled): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::MODULES)
                ->update(["enabled" => $enabled], ["module" => $module]);
        });

        self::register("getModuleState", function (string $module): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::MODULES)
                ->get(["enabled"], ["module" => $module]);
        });

        self::register("enablePlayerNotifications", function (string $player): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::NOTIFICATIONS)
                ->insert(["player" => $player]);
        });

        self::register("disablePlayerNotifications", function (string $player): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::NOTIFICATIONS)
                ->delete(["player" => $player]);
        });

        self::register("hasNotificationsEnabled", function (string $player): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::NOTIFICATIONS)
                ->has(["player" => $player]);
        });

        self::register("addToWhitelist", function (string $player): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::MAINTENANCE_LIST)
                ->insert(["player" => $player]);
        });

        self::register("removeFromWhitelist", function (string $player): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::MAINTENANCE_LIST)
                ->delete(["player" => $player]);
        });

        self::register("isOnWhitelist", function (string $player): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::MAINTENANCE_LIST)
                ->has(["player" => $player]);
        });

        self::register("getWhitelist", function (): QueryBuilder {
            return QueryBuilder::table(DatabaseTables::MAINTENANCE_LIST)
                ->select(["player"], "*");
        });
    }
}

File: src/main/java/de/pocketcloud/cloud/provider/database/DatabaseQueries.java
package de.pocketcloud.cloud.provider.database;

import de.pocketcloud.cloud.template.util.TemplateHelper;

import java.util.Map;

public final class DatabaseQueries {

    public static String createTables() {
        return
            "CREATE TABLE IF NOT EXISTS " + DatabaseTables.TEMPLATES + " (" +
                "name VARCHAR(50) PRIMARY KEY," +
                "lobby BOOL," +
                "maintenance BOOL," +
                "`staticServers` BOOL," +
                "maxPlayerCount INTEGER," +
                "minServerCount INTEGER," +
                "maxServerCount INTEGER," +
                "startNewPercentage FLOAT," +
                "autoStart BOOL," +
                "alwaysCopyToStaticServers BOOL," +
                "templateType VARCHAR(20)" +
                "serverSoftware VARCHAR(30)" +
            "); " +
            "CREATE TABLE IF NOT EXISTS " + DatabaseTables.SERVER_GROUPS + " (" +
                "name VARCHAR(50) PRIMARY KEY," +
                "templates TEXT" +
            "); " +
            "CREATE TABLE IF NOT EXISTS " + DatabaseTables.MODULES + " (" +
                "module VARCHAR(100) PRIMARY KEY," +
                "enabled BOOL" +
            "); " +
            "CREATE TABLE IF NOT EXISTS " + DatabaseTables.NOTIFICATIONS + " (" +
                "player VARCHAR(16) PRIMARY KEY" +
            "); " +
            "CREATE TABLE IF NOT EXISTS " + DatabaseTables.MAINTENANCE_LIST + " (" +
                "player VARCHAR(16) PRIMARY KEY" +
            ");";
    }

    public static String addTemplate() {
        return buildInsert(DatabaseTables.TEMPLATES, TemplateHelper.KEYS.toArray(new String[0]));
    }

    public static String removeTemplate() {
        return "DELETE FROM " + DatabaseTables.TEMPLATES + " WHERE name = ?";
    }

    public static String editTemplate(Map<String, Object> newData) {
        return buildUpdate(DatabaseTables.TEMPLATES, newData, "name");
    }

    public static String getTemplate() {
        return "SELECT " + String.join(", ", TemplateHelper.KEYS) + " FROM " + DatabaseTables.TEMPLATES + " WHERE name = ?";
    }

    public static String checkTemplate() {
        return "SELECT COUNT(*) FROM " + DatabaseTables.TEMPLATES + " WHERE name = ?";
    }

    public static String getTemplates() {
        return "SELECT " + String.join(", ", TemplateHelper.KEYS) + " FROM " + DatabaseTables.TEMPLATES;
    }

    public static String addServerGroup() {
        return buildInsert(DatabaseTables.SERVER_GROUPS, new String[]{"name", "templates"});
    }

    public static String removeServerGroup() {
        return "DELETE FROM " + DatabaseTables.SERVER_GROUPS + " WHERE name = ?";
    }

    public static String editServerGroup(Map<String, Object> newData) {
        return buildUpdate(DatabaseTables.SERVER_GROUPS, newData, "name");
    }

    public static String getServerGroup() {
        return "SELECT name, templates FROM " + DatabaseTables.SERVER_GROUPS + " WHERE name = ?";
    }

    public static String checkServerGroup() {
        return "SELECT COUNT(*) FROM " + DatabaseTables.SERVER_GROUPS + " WHERE name = ?";
    }

    public static String getServerGroups() {
        return "SELECT name, templates FROM " + DatabaseTables.SERVER_GROUPS;
    }

    public static String insertModuleState() {
        return "INSERT INTO " + DatabaseTables.MODULES + " (module, enabled) VALUES (?, ?)";
    }

    public static String setModuleState() {
        return "UPDATE " + DatabaseTables.MODULES + " SET enabled = ? WHERE module = ?";
    }

    public static String checkModuleState() {
        return "SELECT COUNT(*) FROM " + DatabaseTables.MODULES + " WHERE module = ?";
    }

    public static String getModuleState() {
        return "SELECT enabled FROM " + DatabaseTables.MODULES + " WHERE module = ?";
    }

    public static String enablePlayerNotifications() {
        return "INSERT IGNORE INTO " + DatabaseTables.NOTIFICATIONS + " (player) VALUES (?)";
    }

    public static String disablePlayerNotifications() {
        return "DELETE FROM " + DatabaseTables.NOTIFICATIONS + " WHERE player = ?";
    }

    public static String hasNotificationsEnabled() {
        return "SELECT COUNT(*) FROM " + DatabaseTables.NOTIFICATIONS + " WHERE player = ?";
    }

    public static String getNotificationList() {
        return "SELECT player FROM " + DatabaseTables.NOTIFICATIONS;
    }

    public static String addToWhitelist() {
        return "INSERT IGNORE INTO " + DatabaseTables.MAINTENANCE_LIST + " (player) VALUES (?)";
    }

    public static String removeFromWhitelist() {
        return "DELETE FROM " + DatabaseTables.MAINTENANCE_LIST + " WHERE player = ?";
    }

    public static String isOnWhitelist() {
        return "SELECT COUNT(*) FROM " + DatabaseTables.MAINTENANCE_LIST + " WHERE player = ?";
    }

    public static String getWhitelist() {
        return "SELECT player FROM " + DatabaseTables.MAINTENANCE_LIST;
    }

    private static String buildInsert(String table, String[] columns) {
        String cols = String.join(", ", columns);
        String placeholders = "?,".repeat(columns.length);
        placeholders = placeholders.substring(0, placeholders.length() - 1);
        return "INSERT INTO " + table + " (" + cols + ") VALUES (" + placeholders + ")";
    }

    private static String buildUpdate(String table, Map<String, Object> data, String whereKey) {
        StringBuilder sb = new StringBuilder("UPDATE ").append(table).append(" SET ");
        data.keySet().stream()
            .filter(k -> !k.equals(whereKey))
            .forEach(k -> sb.append(k).append(" = ?, "));

        sb.setLength(sb.length() - 2);
        sb.append(" WHERE ").append(whereKey).append(" = ?");
        return sb.toString();
    }
}